### PR TITLE
Native histograms: fix spurios counter reset when merging recoded chunk to normal chunk

### PR DESCRIPTION
**Pull request by krajorama**
### Description of problem

Fix bug in compaction of overlapping chunks:  if you have two identical histogram chunks A and B , the merge over them produces a single chunk C that has the same data as A and B.

Now add one sample to chunk B without creating a new chunk and call it B'. The merge produces a list of samples that contains all common elements plus the new one at the end.

And you would think that the merge produces a single chunk equal to B' , however that's not true in all cases!

 If the extra histogram added to B has a new bucket, then B is [recoded](https://github.com/prometheus/prometheus/blob/71c90c71d41948eea3f33dfd3ee0a5cc581a36f3/tsdb/chunkenc/histogram.go#L566) into B' to contain the new bucket. Which means that iterating over B' yields samples that have en extra empty bucket.

When we merge the chunks A, B' the histogram samples can come from either A or B' randomly(*) and if we get a sample from B' and then from A, then the sample from A is now [not appendable](https://github.com/prometheus/prometheus/blob/71c90c71d41948eea3f33dfd3ee0a5cc581a36f3/tsdb/chunkenc/histogram.go#L304) to the result and in fact causes a new chunk and a **counter reset**!

The real case where I've seen it is a bit more complicated, but the upshot is that we see spikes in the rate/increase of  the data after compaction.

### Solution

Accept this case in appendable: previous histogram has extra empty buckets. This should be appendable and append should just fill in the empty buckets in the incoming histogram.

I ended up making two functions: `expandIntSpansAndBuckets` which is a merge of `counterResetInAnyBucket` and `expandSpansForward`  AND `expandFloatSpansAndBuckets` which is a merge of `counterResetInAnyFloatBucket` and `expandSpansForward` . The merge was pretty straightforward, however it didn't allow for a single generic function. Also the new functions have repeated code, but I couldn't inline it without destroying performance.

### Alternate solution discarded 1

I was thinking that maybe selecting the source chunk wisely in a merge to always choose the sample with most buckets would be a fix, but that's not going to work. Imagine chunk B' from time T1 to T3 that was recoded and has "big" histograms due to a "big" sample at time T3 (big=has more buckets). However for whatever reason it lacks a sample at time T2, which another chunk A has and that sample will be "small" , causing the counter reset on merge. Since B' doesn't have a sample at T2, there's no way to choose the bigger sample.

### Alternate solution discarded 2

I'm not 100% sure we should discard this.

When returning histograms from a chunk iterator, do not return empty buckets. So basically sanitize the buckets. This means adding extra iteration and check on every readout :( 

### Alternate solution discarded 3

Add `h = h.Copy().Compact(0)` before line https://github.com/prometheus/prometheus/blob/71c90c71d41948eea3f33dfd3ee0a5cc581a36f3/storage/series.go#L340.  This would mean making a copy for every histogram. Doubles the compaction time pretty much.

### CODE REVIEW

File: tsdb/chunkenc/histogram_test.go [hunk 1]
```diff
@@ -1368,3 +1447,50 @@ func TestHistogramAppendOnlyErrors(t *testing.T) {
 		require.EqualError(t, err, "histogram counter reset")
 	})
 }
+
+func BenchmarkAppendable(b *testing.B) {
```
Comment (jesusvazquez) [inline tsdb/chunkenc/histogram_test.go:1451]: Can you post these benchmark results?

**Review by jesusvazquez — APPROVED** [review]
LGTM

I did not see anything that was clearly wrong and I appreciate the effort in writing the extensive description of the pull request. The topic is a bit complex and that guided me a bit.

If @fionaliao @carrieedwards or @zenador have a chance to have a look I'd appreciate that as well since they are more familiar with this than I am and also because maybe this also affects the out of order implementation.

**Review by fionaliao — APPROVED** [review]
LGTM

File: tsdb/chunkenc/float_histogram.go [hunk 2]
```diff
@@ -279,27 +289,214 @@ func (a *FloatHistogramAppender) appendable(h *histogram.FloatHistogram) (
...
+	ai := newBucketIterator(a)
+	bi := newBucketIterator(b)
+
+	var aInserts []Insert // To insert into buckets of a, to make up for missing buckets in b.
+	var bInserts []Insert // To insert into buckets of b, to make up for missing empty(!) buckets in a.
+
+	// When aInter.num or bInter.num becomes > 0, this becomes a valid insert that should
+	// be yielded when we finish a streak of new buckets.
+	var aInter Insert
+	var bInter Insert
```
[the remark this answers is not in the document]
Comment (krajorama) [reply tsdb/chunkenc/float_histogram.go:359]: Historical reasons: https://github.com/prometheus/prometheus/blob/7fab72a280f139170a14e6f6a21f6396fa02899e/tsdb/chunkenc/histogram_meta.go#L330

No idea, but I wanted to keep the resemblance to the original

File: tsdb/chunkenc/histogram_test.go (hunk 1 above)
Comment (krajorama) [reply tsdb/chunkenc/histogram_test.go:1451]:
```
             │ /tmp/old.bench │           /tmp/new.bench            │
             │     sec/op     │   sec/op     vs base                │
Appendable-8      173.1µ ± 2%   146.1µ ± 1%  -15.61% (p=0.000 n=10)

             │ /tmp/old.bench │           /tmp/new.bench           │
             │      B/op      │    B/op     vs base                │
Appendable-8       172.0 ± 1%   145.0 ± 1%  -15.70% (p=0.000 n=10)

             │ /tmp/old.bench │         /tmp/new.bench         │
             │   allocs/op    │ allocs/op   vs base            │
Appendable-8       0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```

**Pull request merged**